### PR TITLE
openAPI docs

### PR DESCRIPTION
**Pull request by feep**
This is a draft to render some options on netlify.

Details [here](https://deploy-preview-244--qri-website.netlify.app/docs/api/readme).

Note: the redoc example does not work (for me) when copy-pasting the link, or when reloading. Only when coming from the link in the readme. Gatsby interaction bug, I presume?

### CODE REVIEW

**Comment by chriswhong** [comment]
I think redoc is the most promising.  I looked at your code but you weren't really doing it the react way, which might be why it only loads when navigating from another page on the site.

The right way to load it is with an `import` statement and then use the `<RedocStandalone />` react component:

```
import React from 'react'
import Head from '../components/Head'
import { RedocStandalone } from 'redoc'

const RedocPage = () => (
  <div className='container docs-page'>
    <Head data={{
      title: 'openAPI redoc',
      description: 'todo(rusty)'
    }} />
    <hr/>
    <div className='docs-contents'>
      <RedocStandalone specUrl='/open_api_3_cut_errors.yaml' />
    </div>
  </div>
)

export default RedocPage
```

**Comment by feep** [comment]
I tried that first. Webpack throws an error. I couldn’t (quickly) find where gatsby hides its webpack config, I did what I knew would work (for a sample).

```
 ERROR #98123  WEBPACK
Generating development JavaScript bundle failed
Can't resolve 'redoc' in '/home/rusty/qri/website/src/pages'
```

I’ll figure the react way out when I move it to `docs/`.

**Comment by ramfox** [comment]
I'm assuming we would be adjusting styles etc.

I'm using chrome to view.

rapiddoc: i actually like this layout, it's clean & clear. There is a weird bug where the anchor tags are sometimes not letting me scroll up and down smoothly (if I ever scroll up, it jumps me up to the `#overview` tag which bring me to the start of the page & when I scroll down it will sometimes jump me back up to the previous tag). Also, if you click on an endpoint, for example when I just clicked "profile", the grey sidebar isn't rendered on the side with the rest of the content. If I scroll down, rather then click to a certain anchortag (and I'm not getting that bug I outlined above), then the sidebar is rendered.

redoc: I can see how this would be useful & that having three sections makes all the avail info in one screen. However, and this is probably something we can change, I think the way the sidebar and the sections are delineated are confusing. The only place that mentiones the actual endpoint is in the third section, and that's under a select box. Are we using the 'descriptions' as titles, and can we adjust that to be the endpoints themselves? I think it would be much more clear to me then. Also, the third column should be a much wider portion of the screen, in my opinion :) Those adjustments would make me like redoc way more.

I like the clean of rapiddoc, however, i think if redoc was adjusted (the sections titles & width of the final column) I can see how that would be favorable, since it would bring more info with less scrolling.

I am curious, in redoc, what is the select box supposed to be a selection between? All the ones I've checked on our page have only one option.

**Comment by feep** [comment]
@ramfox 

### tl;dr

- rapidoc
  - I can't see how to fix the scrolling issues without eliminating the toc completely
  - I can make a [wider (3 column-ish) view](https://deploy-preview-244--qri-website.netlify.app/rapidoc-wide) (which does not have the TOC)
- redoc
  - I cannot make the headers the endpoint names without changing the javascript
  - Maybe(?) the column widths can be adjusted with CSS in the three column view
  - I can move the endpoint to column 2

> I'm assuming we would be adjusting styles etc.

Do you have anything in mind besides font?

### rapiddoc

> There is a weird bug where the anchor tags are sometimes not letting me scroll up and down smoothly (if I ever scroll up, it jumps me up to the #overview tag which bring me to the start of the page & when I scroll down it will sometimes jump me back up to the previous tag). 

Oh, I get scrolling issues, too (chromium based) Probably an interaction with something in Gatsby.

I didn't see them when I was playing because I was mostly zoomed in more and it works in single column.

I haven't been able to figure out how to get the scrolling to work in Gatsby, besides setting it to one column (and losing the clickable TOC).

### redoc

> I can see how this would be useful & that having three sections makes all the avail info in one screen. 

Wow, I didn't even see the 3 column layout. My screen is not that wide.

> However, and this is probably something we can change, I think the way the sidebar and the sections are delineated are confusing. The only place that mentions the actual endpoint is in the third section, and that's under a select box.

I can get the endpoint into the middle section.

I cannot figure out how to make the select box disappear. Even with only one server like I have in the demo `.yaml`.

> Are we using the 'descriptions' as titles, and can we adjust that to be the endpoints themselves? 

I do not see a way to do this without editing the javascript.

> I think it would be much more clear to me then. Also, the third column should be a much wider portion of the screen, in my opinion :) Those adjustments would make me like redoc way more.

Maybe (hopefully) the third column can be widened with css.

> I am curious, in redoc, what is the select box supposed to be a selection between? All the ones I've checked on our page have only one option.

Looks like a choice between servers. And there is only one in the `.yaml` I am using.

There are two server choices [here](https://redocly.github.io/redoc/#operation/addPet).

**Comment by ramfox** [comment]
I can't see the redoc version in this latest preview:

```
Something went wrong...
Cannot read property 'style' of undefined
Stack trace
TypeError: Cannot read property 'style' of undefined
    at new e (https://cdn.jsdelivr.net/npm/redoc/bundles/redoc.standalone.js:103:52461)
    at https://cdn.jsdelivr.net/npm/redoc/bundles/redoc.standalone.js:103:59611
    at Array.map (<anonymous>)
    at e.get (https://cdn.jsdelivr.net/npm/redoc/bundles/redoc.standalone.js:103:59587)
    at e.get (https://cdn.jsdelivr.net/npm/redoc/bundles/redoc.standalone.js:103:39843)
    at https://cdn.jsdelivr.net/npm/redoc/bundles/redoc.standalone.js:109:120027
    at gs (https://cdn.jsdelivr.net/npm/redoc/bundles/redoc.standalone.js:79:107583)
    at cl (https://cdn.jsdelivr.net/npm/redoc/bundles/redoc.standalone.js:79:96717)
    at sl (https://cdn.jsdelivr.net/npm/redoc/bundles/redoc.standalone.js:79:96642)
    at Zs (https://cdn.jsdelivr.net/npm/redoc/bundles/redoc.standalone.js:79:93672)

ReDoc Version: 2.0.0-rc.37
Commit: 8e5a5e8
```

I actually like this rapiddoc version, it's super clean looking. I'm wondering how much we lose by not having a search bar? Thoughts? I am also concerned about having a "Try" button that doesn't work. Do we think that's something that can be set up when we go live? If not, I don't think we can use rapiddoc, the "Try" button is to central to the design. 

>Do you have anything in mind besides font?
Yes! For redoc, I was mainly thinking about color.

>I do not see a way to do this without editing the javascript.
Ahh, okay then we will have to edit our descriptions to be more concise in order to truly make sense with this layout. I can see from the example redoc page how they intended it to be used.

>I can get the endpoint into the middle section.
Dope

**Comment by feep** [comment]
> I can't see the redoc version in this latest preview:

Sorry, I'll repush it and let you know. But nothing changed.

> I actually like this rapiddoc version, it's super clean looking. 

I prefer it.

And the lack of navigation makes it less likely to conflict with Gatsby.

> I'm wondering how much we lose by not having a search bar? Thoughts? 

Me? I do not tacked-on search stuff for a single page. If browser find works, I almost always prefer that.

But I'm weird.

And the search box can find things that hidden by being collapsed.

> I am also concerned about having a "Try" button that doesn't work.

Try button can be turned off.

Only reason it is on is to produce the `curl` examples.

> Do we think that's something that can be set up when we go live? If not, I don't think we can use rapiddoc, the "Try" button is to central to the design.

We can turn it off. And turn it on if we want to demo it when there is a public openAPI endpoint.

> Yes! For redoc, I was mainly thinking about color.

The orange and blue highlights? Or the boxes around the http verbs?

**Comment by feep** [comment]
OK, pushed again.

- redoc fixed (fatfingered something in the `.yaml`)
- removed 'try' button from rapidoc-wide

If you have no other things you want me to try, @ramfox & @chriswhong, can you two come to an agreement on [redoc](https://deploy-preview-244--qri-website.netlify.app/docs/api/readme#redoc) (you have to get to it from the readme or it won't load the spec. I _hope_ using the react component will fix that) or [rapidoc-wide](https://deploy-preview-244--qri-website.netlify.app/rapidoc-wide)?

Then I'll push a real page.

I prefer rapidoc-wide.

Some comments:

- I am pretty sure I'll have to write a gatsby plug-in in order to set settings on redoc. Maybe Chris knows a way around this.
- Both will probably be a pain go get to render under `/docs`, but I know no way around this
- I didn't realize this when I first rigged this up, it's nowhere on the github page or demo page. Redoc.js is written by a saas company, and you can see [here](https://redoc.ly/docs/cli/configuration/reference-docs/) that they are putting some new features under the premium label. Not an issue now, could get weird in the future.

**Comment by feep** [comment]
I have an [openAPI page](https://deploy-preview-244--qri-website.netlify.app/docs/api/openapi).

Have some questions on css, `.jsx` weirdness and other.

0. Is there a preferred/proper way to bring a .js (non-react, non-gatsby) dependency into the gatsby site?
I can `yarn add` it, but I don't know how to get it from there to somewhere that my react component can see it.

0. I've poked around in the inspector. I cannot figure why the <rapi-doc> element is 15% narrower than everything else in mainwrapper. 
I can see no padding or margin (besides the 32px below). Something I don't understand is making it narrow. Help?

0. I can gain 32 more px by removing padding from `.main-content-inner--view-mode` but I am unable to write a selector to change the style.
I _think_ this is due to jsx breaking kebab-case, but because there are two hypens in the classname, I am unable to select with a lumpyCase version of the name.

...and any other comments are appreciated.

File: src/components/Rapidoc.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import React from 'react'
+import Helmet from 'react-helmet'
+
+const Rapidoc = () => {
+  return (
+    <div className='container docs-page'>
+      <Helmet>
+        <script type="module" src="https://unpkg.com/rapidoc/dist/rapidoc-min.js"></script>
```
Comment (feep) [inline src/components/Rapidoc.js:8]: I need to know where/how to put this `.js` in the tree.

Drop it in static and commit? Some yarn/gatsby thing I do not know?

File: src/components/Rapidoc.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        <style type="text/css">{`
+        .main-content-inner--view-mode{
```
Comment (feep) [inline src/components/Rapidoc.js:10]: I am unable to write a working selector for this class. I think it is `.jsx` kebab-case weirdness.

**Comment by uhLeeshUh** [comment]
@feep so off the bat, it seems like the problem children are:

- the `.container` class which is adding 15px of padding left & right (specifically the div with `.container docs-page`)
- the `section-gap` class is adding 24px of padding to left & right as well (this applies to all divs with this class)

To your question about not being able to actually access the css selector for `.main-content-inner--view-mode`, that is unfortunately beyond my jsx understanding and I'll defer to @chriswhong or others. I did some initial googling but didn't find anything promising

File: src/components/Rapidoc.js (hunk 1 above)
Comment (uhLeeshUh) [reply src/components/Rapidoc.js:8]: Will also defer to others here, but could you install the rapidoc dep through yarn and import it into this React component? I think I found a good example [here](https://codesandbox.io/s/upbeat-sea-4tfo7?file=/src/App.js)

**Comment by feep** [comment]
> * the `.container` class which is adding 15px of padding left & right (specifically the div with `.container docs-page`)

Oh, thanks, that’s it.

Not the padding (which is minimal), but the margin of `auto`. That's what is eating up the space.

> * as you said, the `.main-content-inner--view-mode` class is adding 16px of padding left & right

Got most of the missing space by fixing the margin `auto`. Not going to bang my head against the wall for the additional 32px unless it’s important to someone.

File: src/components/Rapidoc.js (hunk 1 above)
Comment (feep) [reply src/components/Rapidoc.js]: I pulled `https://unpkg.com/rapidoc/dist/rapidoc-min.js` into `/static/js`.

File: src/components/Rapidoc.js (hunk 2 above)
Comment (feep) [reply src/components/Rapidoc.js]: Going to ignore this. If anybody knows how to gain the additional 32px, I’d love to push a change.

File: src/components/Rapidoc.js (hunk 1 above)
Comment (feep) [reply src/components/Rapidoc.js]: Sorry, your comment didn’t display before I replied to myself.

No idea. Looks like the right way to do it, I’ll give it a shot.

**Comment by feep** [comment]
I think this does everything we want.

Alicia’s react example looks like the proper way to do it. But I blew up gatsby again trying it.

This PR works. If it looks good we can merge.

I can make an issue to convert to react once I get gatsby running again.

File: content/docs/api/openapi.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
+---
+metaTitle: 'Qri OpenAPI'
+metaDescription: 'Documentation for the Qri OpenAPI'
+weight: 1
+
```
Comment (b5) [inline content/docs/api/openapi.md:5]: We haven't finished updating the _actual_ documentation, correct? If so, is there any way to mark this page as a draft until we're done updating the doc? I don't want to publish out of date docs, it'll only make things worse

Comment (feep) [reply content/docs/api/openapi.md:5]: I do not know about the status of the `.yaml` documentation.

There is an issue for errors in the spec. qri-io/qri#1482

But the correctness compared to the actual served API, I do not know.

Note it as as a draft in the text of the page? Or do you mean something else?

**Comment by feep** [comment]
Thanks @b5.

I think that's all, I just need clarification on what you mean by 'mark as draft'.

**Comment by chriswhong** [comment]
Closing in favor of #256
